File: algorithms/ar-radsrt/cpp/radix_sort.cpp
#include<iostream>
#include<list>
#include<cmath>
using namespace std;
void display(int *array, int size) {
   for(int i = 0; i<size; i++)
      cout << array[i] << " ";
   cout << endl;
}
void radixSort(int *arr, int n, int max) {

   list<int> pocket[10];      //radix of decimal number is 10
   for(int i = 0; i< max; i++) {
      int m = pow(10, i+1);
      int p = pow(10, i);
      for(int j = 0; j<n; j++) {
         int temp = arr[j]%m;
         int index = temp/p;      //find index for pocket array
         pocket[index].push_back(arr[j]);
      }
      int count = 0;
      for(int j = 0; j<10; j++) {
         //delete from linked lists and store to array
         while(!pocket[j].empty()) {
            arr[count] = *(pocket[j].begin());
            pocket[j].erase(pocket[j].begin());
            count++;
         }
      }
   }
}
int main() {
   int n, max;
   cout << "Enter the number of elements: ";
   cin >> n;
   cout << "Enter the maximum digit of elements: ";
   cin >> max;
   int arr[n]; //create an array with given number of elements
   cout << "Enter elements:" << endl;
   for(int i = 0; i<n; i++) {
      cin >> arr[i];
   }
   cout << "Data before Sorting: ";
   display(arr, n);
   radixSort(arr, n, max);
   cout << "Data after Sorting: ";
   display(arr, n);
}

File: algorithms/ar-radsrt/python2/radix_sort.py
# Function to get the maximum value in the list
def getMax(arr):
    max_val = arr[0]
    for i in arr:
        if i > max_val:
            max_val = i
    return max_val

# Counting sort to sort elements based on the significant digit
def countingSort(arr, exp):
    n = len(arr)
    output = [0] * n  # Output array to hold sorted elements
    count = [0] * 10  # Initialize counting array (for base 10 digits)
    
    # Store the count of occurrences
    for i in range(n):
        index = (arr[i] // exp) % 10
        count[index] += 1

    # Change count[i] so that count[i] contains the actual position
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Build the output array
    i = n - 1
    while i >= 0:
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1
        i -= 1

    # Copy the output array to arr
    for i in range(n):
        arr[i] = output[i]

# Main function to implement Radix Sort
def radixSort(arr):
    # Find the maximum number to determine the number of digits
    max_val = getMax(arr)
    # Do counting sort for every digit
    exp = 1
    while max_val // exp > 0:
        countingSort(arr, exp)
        exp *= 10

# Example usage
arr = [170, 45, 75, 90, 802, 24, 2, 66]
print("Original array:", arr)
radixSort(arr)
print("Sorted array:", arr)

